OutboxSample/Controllers/ApplicationControllerBase.cs

```diff
@@ -1,14 +1,39 @@
 ﻿using Microsoft.AspNetCore.Mvc;
 using OutboxSample.Application;
+using OutboxSample.Model.Events;
 
 namespace OutboxSample.Controllers;
 
 public abstract class ApplicationControllerBase : ControllerBase
 {
-    protected IUnitOfWorkFactory UnitOfWork { get; }
+    protected IEventMetadataProvider EventMetadataProvider { get; }
+    protected ITimeProvider TimeProvider { get; }
 
-    protected ApplicationControllerBase(IUnitOfWorkFactory unitOfWorkFactory)
+    protected ApplicationControllerBase(IEventMetadataProvider eventMetadataProvider, ITimeProvider timeProvider)
     {
-        this.UnitOfWork = unitOfWorkFactory;
+        this.EventMetadataProvider = eventMetadataProvider;
+        this.TimeProvider = timeProvider;
+    }
+
+    protected EventEnvelope WrapEvent<TEvent>(
```

Comment:
Should the controller be responsible for publishing an event, or would that best be moved to a service or repository layer? What if events need to published that are created outside of an HTTP request?

Reply:
Good point. Publishing on repo layer - certainly no.
Later I'm planning to separate the presentation layer (controller) and application layer (I'm thinking either service or command/query handlers). So, publishing events will be the responsibility of the application layer.